### ISSUE TITLE
Merging 2 APNGs

### ISSUE THREAD

**Issue by SoniEx2** [apngasm/apngasm#50]
apngasm should do a "fast-merge" operation to merge 2 APNGs, in which it only checks the first frame of the next PNG and attempts to optimize only that.

This would let me run 4 apngasm in parallel and then combine the result, as an alternative to #48.

**Comment by Kagetsuki** [comment]
Are you suggesting some way to disable cross-frame comparisons and full-file optimization?

**Comment by SoniEx2** [comment]
Not cross-frame comparisons, but maybe full-file optimization. I mean, if you have 2 assembled APNGs, and you assume they're already optimized, you could just fast-merge them by optimizing the first frame and only the first frame of the second APNG.

e.g.

    [APNG 1 frame 1, copied as-is][APNG 1 frames, copied as-is...][APNG 1 frame n, copied as-is][APNG 2 frame 1, optimized for previous frames][APNG 2 frame 2, copied as-is][APNG 2 frames, copied as-is...][APNG 2 frame n, copied as-is]

instead of re-optimizing everything.

(I guess this is more of a double request: the ability to directly combine 2 APNGs, and make it fast)

**Comment by Kagetsuki** [comment]
Aaah, but the thing is there's full file compression and that relies on data between frames. To open or take apart the file you need to decompress it. Really that's all that's happening there so what you're thinking wouldn't really be possible unless there was no compression used to begin with.

Basically think of an APNG file as a zip file or compressed tarball - you can open it up and take out each of the files/frames but just in the process of opening it up you are decompressing it. Then after you take out all the files and put them together you need to re-compress it.

**Comment by SoniEx2** [comment]
Uhh what do you mean, frames aren't compressed individually?

**Comment by Kagetsuki** [comment]
> because even if it is the same as the previous frame, it's only up to about double the size

Actually it's checking if any other frame is the same as any other frame; THEN when all the frames are packed together it compresses them all together which ends up doing things like looking for and reducing repetitive data between frames. Actual in-frame optimization is probably already done during the initial encoding to PNG which is likely done before apngasm is even involved.

The tool is making a best effort optimization. You're free to put up an MR to make this optimization optional but it will involve deconstructing the optimization code - a lot of which is carried over from apngasm 1~/2~ and it's a bit of a rats nest. On top of that cluttering the code base and making invocation more complex with options 99% of developers/users won't want isn't something I think me or any of the apngasm maintainers would want.

> I mean, unless the APNG has gamma correction and stuff, you should be able to just merge them

Sort of. While technically you can just smash together frames and edit the headers there are a variety of end user issues with that. We want to make the best file possible with as little hassle as possible and that requires doing as many optimization tricks as we can - even if they take time. APNG is a relatively data-heavy format so we really need to make an effort to reduce any weight we can while also keeping the format quick to read and easy to handle. That means that assembly/encoding/compilation of files is a heavy process simply because it needs to be.

The reality of the situation is you're trying to do something apngasm wasn't meant to do and in the end you're going to end up with a huge and very non-portable file. To encode it your best option is to just put all the frames as png files in a folder, set the tool to work, and wait for it to finish. I have a feeling though you will be disappointed with the result.

**Comment by SoniEx2** [comment]
You don't need to recompress each previous frame because they already are compressed. If anything you can just decompress each previous frame and optimize the first frame of the second APNG for the previous frames. The thing is, even if you do decompress each previous frame you're still saving a lot of time as you won't be wasting time recompressing *everything*.

**Comment by Kagetsuki** [comment]
Efficient compression requires you compress all data together. You will not get efficient compression if you just compress each frame and then pack them together.

As an experiment: take 4 PNG files and add up their size separately, then turn then into an APNG and look at the size of the file. The APNG file should be smaller because it's compressing data between frames, not just in each frame separately.

To further illustrate this take the example of compressed tarballs. If you tar a bunch of files and then compress them you will get a file smaller than if you compressed a bunch of files and then tarred them.

**Comment by SoniEx2** [comment]
Weird, when I tried it the APNG was bigger.

Maybe you should read your own code?

**Comment by Kagetsuki** [comment]
I've just re-tested and I can't find an example where the result file was larger than the sum of the source frame files. Can you upload the files you tested with somewhere so I can try to replicate your results?

**Comment by SoniEx2** [comment]
Use it for a slideshow, instead of an animation.

**Comment by Kagetsuki** [comment]
Do you mean to increase the delay? If so I found no increase in file size.

**Comment by SoniEx2** [comment]
No I mean use various JPEG images re-encoded as PNG.

**Comment by Kagetsuki** [comment]
PNG uses an entirely different encoding scheme and has at least one additional bit depth for alpha and a different encoding and optimization scheme. In general a large image such a a photo or a frame of video without any alpha will have a greater data weight in PNG than JPEG (the opposite is often true for things like animation sprites). In addition to that apngasm does not modify or override any base PNG encoding schemes - it's all entirely standard libpng encoding; so there is effectively nothing we can do to reduce per-frame data weight within apngasm without breaking compliance and possibly impairing compatibility.

It should be noted that the added depth and broad array of features is what makes PNG an optimal choice for an animated image format - but it is also this added depth that makes it less preferable as a video format.

**Comment by SoniEx2** [comment]
Then take multiple screenshots of Minecraft and combine them into a slideshow. (Minecraft outputs PNGs anyway.)

Anyway my point is, APNG doesn't do cross-frame *compression* but cross-frame *elimination* (it removes the parts that didn't change - see `fcTL`'s `blend_op` and `dispose_op`), and according to that it's a simple matter of changing all the frame IDs. (the only thing that would require complete recompression is `APNG_DISPOSE_OP_PREVIOUS`-related optimizations)

**Comment by Kagetsuki** [comment]
The elimination you're mentioning is part of the compression scheme. If you think you can do it more efficiently we'd be happy to review a patch/merge request!

**Comment by galaxy001** [comment]
I just find [gif2apng](http://gif2apng.sourceforge.net/) support `APNG_DISPOSE_OP_PREVIOUS`, and all two `blend_op`s.
